src/include/misc.h

```diff
@@ -274,10 +274,24 @@
  */
 #define WT_STREQ(s, cs) (sizeof(cs) == 2 ? (s)[0] == (cs)[0] && (s)[1] == '\0' : strcmp(s, cs) == 0)
 
-/* Check if a string matches a byte string of len bytes. */
-#define WT_STRING_MATCH(str, bytes, len)                                                        \
-    (((const char *)(str))[0] == ((const char *)(bytes))[0] && strncmp(str, bytes, len) == 0 && \
-      (str)[len] == '\0')
+/*
+ * Check if a string matches a byte string of len bytes. This macro expects both strings to be of
+ * non-zero length in order to be a match - this means that strings cannot be a match if len
+ * == 0.
+ *
+ * Use an inline function as that tends to be safer and we can check for NULL values.
+ */
+#define WT_STRING_MATCH(str, bytes, len) __wt_string_match(str, bytes, len)
+
+static inline bool
+__wt_string_match(const char *str, const char *bytes, size_t len)
+{
+    if (str != NULL && bytes != NULL && len > 0 && (strlen(str) == len) &&
```

Comment:
As discussed I think the null checks should be asserts: as I think of this function as having the same kind of caller contract as any of the `str*` library functions where NULL is not valid string.

Reply:
Agreed. I will mostly move these checks into an assert statement to be executed in diagnostic mode.

Reply:
Actually I think I will remove it altogether, it seems a bit clunky to pass in the session to do this assert given that most string functions do not do this check anyways.